lib/umwelt/client/clone.ex

```diff
@@ -0,0 +1,95 @@
+defmodule Umwelt.Client.Clone do
+  @moduledoc "Clone main process"
+
+  use GenServer
+  require Logger
+
+  alias Umwelt.Client
+
+  def start_link(_),
+    do:
+      GenServer.start_link(
+        __MODULE__,
+        %{phase_id: nil, port: nil},
+        name: __MODULE__
+      )
+
+  def init(state), do: {:ok, state}
+
+  def handle_cast({:pull, params}, _state) do
+    send(self(), :start_pulling)
+
+    {:noreply, params}
+  end
+
+  def handle_info(:start_pulling, state) do
+    case Client.Request.fetch_modules(state) do
+      {:ok, modules} ->
+        Logger.info("Fetching modules: #{inspect(Map.keys(modules))}")
+        modules |> Client.Agent.add_modules()
+
+      {:error, reason} ->
+        Logger.error("Failed to fetch modules: #{inspect(reason)}. Stopping...")
+        Supervisor.stop(Client.Supervisor)
+    end
+
+    send(self(), :spawn_fetchers)
+
+    {:noreply, state}
+  end
+
+  def handle_info(:spawn_fetchers, state) do
+    total = Client.Agent.total()
+    Logger.debug("Spawning fetchers: #{total}")
+
+    Client.Agent.all_waiting()
+    |> Enum.each(fn _ -> spawn_fetcher(state) end)
+
+    {:noreply, state}
+  end
+
+  def handle_info({:fetched, %{name: name, code: code}}, state) do
+    Client.Agent.update_status(name, :fetched)
+    spawn_writer(%{name: name, code: code})
+    {:noreply, state}
+  end
+
+  def handle_info({:fetch_failed, module}, state) do
+    Logger.warning("Respawning failed fetcher for module #{module.name}")
+    Client.Fetcher.start_link(module)
+    {:noreply, state}
+  end
+
+  def handle_info({:written, mod_name}, state) do
+    Client.Agent.update_status(mod_name, :written)
+    send(self(), :maybe_stop)
+    {:noreply, state}
+  end
+
+  def handle_info(:maybe_stop, state) do
+    if Client.Agent.completed?() do
+      Logger.debug("All modules processed. Stopping application.")
+      Supervisor.stop(Client.Supervisor)
+    end
+
+    {:noreply, state}
+  end
+
+  defp spawn_fetcher(state) do
+    case Client.Agent.next_waiting() do
+      nil ->
+        Logger.debug("No more modules to fetch")
+        :ok
+
+      module ->
+        Logger.debug("Spawning fetcher for module #{inspect(module.name)}")
+        Client.Fetcher.start_link(Map.merge(module, state))
+    end
+  end
+
+  defp spawn_writer(%{name: name} = module) do
+    Logger.debug("Spawning writer for module #{inspect(name)}")
+    Client.Agent.update_status(name, :writing)
+    Client.Writer.start_link(module)
```

Comment:
I would better organize writers under their own `DynamicSupervisor` rather than linking them all to this process. One failed writer should be restarted, it should not crash the whole process.

Reply:
Maybe `Task.Supervisor` because workers are tasks?